### PR TITLE
quick configure buttons for management console

### PR DESCRIPTION
**Pull request by beyang**
Add "Quick configure" buttons to quickly configure auth, licenses, and external URL in critical site config:

image

Fixes https://github.com/sourcegraph/sourcegraph/issues/3218

- [ ] Code review: @slimsag 
- [ ] Design/product review: @francisschmaltz @sqs

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -18,6 +18,7 @@ All notable changes to Sourcegraph are documented in this file.
 - The new `bitbucketserver.repos` setting in [Bitbucket Server external service config](https://docs.sourcegraph.com/admin/external_service/bitbucketserver#configuration) allows you to select specific repositories to be synced.
 - The new required `bitbucketserver.repositoryQuery` setting in [Bitbucket Server external service configuration](https://docs.sourcegraph.com/admin/external_service/bitbucketserver#configuration) allows you to use Bitbucket API repository search queries to select repos to be synced. Existing configurations will be migrate to have it set to `["?visibility=public", "?visibility=private"]` which is equivalent to the previous implicit behaviour that this setting supersedes.
 - "Quick configure" buttons for common actions have been added to the config editor for all external services.
+- "Quick configure" buttons for common actions have been added to the critical site config editor in the management console.
```
Comment (slimsag) [inline CHANGELOG.md]:
```suggestion
- "Quick configure" buttons for common actions have been added to the management console.
```

File: cmd/management-console/web/src/CriticalConfigEditor.tsx [hunk 2]
```diff
@@ -129,40 +330,128 @@ export class CriticalConfigEditor extends React.Component<Props, State> {
...
-                    {this.state.criticalConfig && (
-                        <MonacoEditor
-                            content={this.state.criticalConfig.Contents}
-                            language="json"
-                            onDidContentChange={this.onDidContentChange}
-                            onDidSave={this.onDidSave}
-                        />
+            <div>
+                {actions && (
+                    <div className="critical-config-page__action-groups">
```
Comment (slimsag) [inline cmd/management-console/web/src/CriticalConfigEditor.tsx]: It is fine if these are in the editor, but the editor is a component not a page, so naming it `critical-config-page` is misleading. The proper BEM name would be `critical-config-editor`, so you should be using just that.

File: cmd/management-console/web/src/CriticalConfigEditor.tsx [hunk 3]
```diff
@@ -129,40 +330,128 @@ export class CriticalConfigEditor extends React.Component<Props, State> {
...
-                >
-                    {!this.state.criticalConfig && this.state.canShowLoader && <div>Loading...</div>}
-                    {this.state.criticalConfig && (
-                        <MonacoEditor
-                            content={this.state.criticalConfig.Contents}
-                            language="json"
-                            onDidContentChange={this.onDidContentChange}
-                            onDidSave={this.onDidSave}
-                        />
+            <div>
```
Comment (slimsag) [inline cmd/management-console/web/src/CriticalConfigEditor.tsx]: Do we actually need this div? Can't the actions be put under `<div className="critical-config-editor">` below?

File: cmd/management-console/web/src/CriticalConfigEditor.tsx [hunk 4]
```diff
@@ -129,40 +330,128 @@ export class CriticalConfigEditor extends React.Component<Props, State> {
         this.subscriptions.unsubscribe()
     }
 
+    private runAction(id: string, editor?: _monaco.editor.ICodeEditor): void {
+        if (editor) {
+            const action = editor.getAction(id)
+            action.run().then(() => void 0, (err: any) => console.error(err))
+        } else {
+            alert('Wait for editor to load before running action.')
```
Comment (slimsag) [inline cmd/management-console/web/src/CriticalConfigEditor.tsx]: This is a bad UX. Ideally the buttons would be disabled until the editor loaded.

Un lieu of that, just don't alert. i.e. remove the else case and add a comment explaining why `if (editor)` exists.

File: cmd/management-console/web/src/CriticalConfigEditor.tsx [hunk 5]
```diff
@@ -129,40 +330,128 @@ export class CriticalConfigEditor extends React.Component<Props, State> {
...
+        }
+    }
+
+    public static isStandaloneCodeEditor(
```
Comment (slimsag) [inline cmd/management-console/web/src/CriticalConfigEditor.tsx]: As far as I understand, this just tells us if the editor is a code editor or diff editor, right?

Why is this named `isStandaloneCodeEditor` then? Why not `isCodeEditor`?

File: enterprise/cmd/frontend/auth/saml/config.go [hunk 6]
```diff
@@ -50,13 +50,13 @@ func handleGetProvider(ctx context.Context, w http.ResponseWriter, pcID string)
...
-		log15.Error("No SAML auth provider found with ID.", "id", pcID)
+		log15.Error("No SAML auth provider found with ID", "id", pcID)
 		http.Error(w, "Misconfigured SAML auth provider.", http.StatusInternalServerError)
 		return nil, true
 	}
 	if err := p.Refresh(ctx); err != nil {
-		log15.Error("Error refreshing SAML auth provider.", "id", p.ConfigID(), "error", err)
-		http.Error(w, "Unexpected error refreshing SAML authentication provider.", http.StatusInternalServerError)
+		log15.Error("Error getting SAML auth provider", "id", p.ConfigID(), "error", err)
+		http.Error(w, "Unexpected error getting SAML authentication provider. This may indicate that the SAML IdP does not exist. Ask a site admin to check the server \"frontend\" logs for \"Error getting SAML auth provider\".", http.StatusInternalServerError)
```
Comment (slimsag) [inline enterprise/cmd/frontend/auth/saml/config.go]: This change shouldn't be here?

File: cmd/management-console/web/src/CriticalConfigEditor.tsx [hunk 7]
```diff
@@ -222,3 +511,67 @@ export class CriticalConfigEditor extends React.Component<Props, State> {
...
+    try {
+        externalURL = jsonc.parse(config).externalURL
+        externalURLRegexp = externalURL.replace(/\//g, '\\/')
+    } catch {
+        /* not necessarily an error, config might be empty */
+    }
+    if (!externalURL) {
+        externalURL = '<externalURL>'
+        externalURLRegexp = '<externalURL regex>'
+    }
```
Comment (slimsag) [inline cmd/management-console/web/src/CriticalConfigEditor.tsx]: Instead of using this, can't we just set these defaults in the `catch` block?

Comment (slimsag) [reply cmd/management-console/web/src/CriticalConfigEditor.tsx]: Actually, can you just write:

```typescript
return { externalURL, externalURLRegexp }
```

In both blocks? That would be a lot cleaner

File: cmd/management-console/web/src/CriticalConfigEditor.tsx [hunk 8]
```diff
@@ -222,3 +511,67 @@ export class CriticalConfigEditor extends React.Component<Props, State> {
...
+        range: _monaco.Range.fromPositions(
+            model.getPositionAt(edit.offset),
+            model.getPositionAt(edit.offset + edit.length)
+        ),
+        forceMoveMarkers: true,
+        text: edit.content,
+    }))
+}
+
+function getPositionAt(text: string, offset: number): _monaco.IPosition {
```
Comment (slimsag) [inline cmd/management-console/web/src/CriticalConfigEditor.tsx]: Document what `offset` is? Byte offset? Character offset?

File: cmd/management-console/web/src/CriticalConfigEditor.tsx [hunk 9]
```diff
@@ -129,40 +330,128 @@ export class CriticalConfigEditor extends React.Component<Props, State> {
...
+    private editorWillMount = (editor: _monaco.editor.IStandaloneCodeEditor, model: _monaco.editor.IModel) => {
+        this.configEditor = editor
+
+        if (CriticalConfigEditor.isStandaloneCodeEditor(editor)) {
+            for (const { id, label, run } of quickConfigureActions) {
+                editor.addAction({
+                    label,
+                    id,
+                    run: editor => {
+                        // copy-pasta
```
Comment (slimsag) [inline cmd/management-console/web/src/CriticalConfigEditor.tsx]: From where? Link to it at least?

Comment (beyang) [reply cmd/management-console/web/src/CriticalConfigEditor.tsx]: It was originally copy-pasted, but the code has diverged enough now that this comment is no longer necessary. I've removed the comment.

File: enterprise/cmd/frontend/auth/saml/config.go (hunk 6 above)
Comment (beyang) [reply enterprise/cmd/frontend/auth/saml/config.go]: this was a related change to address https://github.com/sourcegraph/sourcegraph/issues/3218

File: cmd/management-console/web/src/CriticalConfigEditor.scss [hunk 10]
```diff
@@ -1,7 +1,6 @@
 .critical-config-editor {
     &__monaco-reserved-space {
-        width: 45rem;
-        height: 32rem;
+        height: 50rem;
```
Comment (slimsag) [inline cmd/management-console/web/src/CriticalConfigEditor.scss]: My single biggest complaint is this change here.

This is how it looks on my 15in Retina macbook screen:

image

Worse, I have to scroll the entire page to get to the save buttons. I think this is a very common screen size, so I am opposed to this change. I would rather the size remain the same and the buttons e.g. wrap around.

Comment (beyang) [reply cmd/management-console/web/src/CriticalConfigEditor.scss]: Reverted to original height/width and made editor expandable.

File: cmd/management-console/web/src/CriticalConfigEditor.tsx (hunk 7 above)
Comment (beyang) [reply cmd/management-console/web/src/CriticalConfigEditor.tsx]: No, because `externalURL` might be `undefined` without triggering an exception (e.g., if the JSON parses properly, but doesn't have an `externalURL` field).

**Comment by beyang** [comment]
@slimsag Merged, because screen size issue resolved (reverted to old size, made editor resizable).

File: cmd/management-console/web/src/CriticalConfigEditor.tsx [hunk 11]
```diff
@@ -129,21 +333,56 @@ export class CriticalConfigEditor extends React.Component<Props, State> {
         this.subscriptions.unsubscribe()
     }
 
+    private runAction(id: string, editor?: _monaco.editor.ICodeEditor): void {
+        if (editor) {
+            const action = editor.getAction(id)
+            action.run().then(() => void 0, (err: any) => console.error(err))
```
Comment (felixfbecker) [inline cmd/management-console/web/src/CriticalConfigEditor.tsx]: Use `.catch()` instead of `.then(() => void 0`

File: cmd/management-console/web/src/CriticalConfigEditor.tsx [hunk 12]
```diff
@@ -1,3 +1,7 @@
+import * as jsonc from '@sqs/jsonc-parser'
+import { setProperty } from '@sqs/jsonc-parser/lib/edit'
+import * as _monaco from 'monaco-editor'
```
Comment (felixfbecker) [inline cmd/management-console/web/src/CriticalConfigEditor.tsx]: The underscore indicates that this import should not be referenced except for types (for lazy loading), but it is referenced for values in this file

File: cmd/management-console/web/src/CriticalConfigEditor.scss [hunk 13]
```diff
@@ -1,7 +1,12 @@
 .critical-config-editor {
+    width: 50rem;
+    margin-bottom: 10rem;
+
     &__monaco-reserved-space {
-        width: 45rem;
-        height: 32rem;
+        height: 35rem;
+        overflow: hidden;
+        resize: both;
```
Comment (felixfbecker) [inline cmd/management-console/web/src/CriticalConfigEditor.scss]: Doesn't this look funky if you extend it beyond the container width? I think I would go for only Y axis resizing

File: cmd/management-console/web/src/CriticalConfigEditor.tsx (hunk 5 above)
Comment (felixfbecker) [reply cmd/management-console/web/src/CriticalConfigEditor.tsx]: Wondering the same

File: web/src/settings/MonacoSettingsEditor.tsx [hunk 14]
```diff
@@ -208,7 +208,8 @@ export class MonacoSettingsEditor extends React.PureComponent<Props, State> {
     public static isStandaloneCodeEditor(
         editor: monaco.editor.ICodeEditor
     ): editor is monaco.editor.IStandaloneCodeEditor {
-        return editor.getEditorType() === monaco.editor.EditorType.ICodeEditor
+        const editorAny = editor as any
```
Comment (felixfbecker) [inline web/src/settings/MonacoSettingsEditor.tsx]: Why is this needed, and why is this implementation different than the other `isStandaloneCodeEditor()`? Why are there even two implementations?
